Run lazy tensor codegen in generate_code.py

wconstab opened the pull request:
Summary:
Instead of gen.py, run in generate_code / autograd generator which is more native to
torch/csrc rather than aten

Differential Revision: D34408536

ezyang commented on aten/src/ATen/templates/LazyIr.h:
```diff
@@ -0,0 +1,21 @@
+
+namespace torch {
+namespace lazy {
+using at::operator<<;
```
this is probably cargo culted but could you file an issue about making sure ADL actually works for `operator<<` in all cases. This is probably a case of someone sticking an `operator<<` overload in the wrong namespace

wconstab replied:
It might be possible that we only had this issue back when LazyIr.h was using its own namespace (not torch::lazy).  I will check if it can be removed now.  But you might still want me to file that issue since at least it was a problem in our old namespace setup.

ezyang commented on aten/src/ATen/templates/LazyIr.h:
```diff
@@ -0,0 +1,21 @@
+// to differentiate between HASH(nullopt, something) and HASH(something, nullopt),
+// and using kNullValue in the hash function in the order of arguments
+// serves this purpose.
+static const torch::lazy::Value kNullValue = torch::lazy::Value();
```
static value in header file? This is my suspicious face.

wconstab replied:
oh yea, this looks really dumb.  I'll figure out why we did this and move it somewhere better..

wconstab replied:
oh.. i remember now. This isn't too bad since all I was going for was to not call the Value() constructor a ton of times, but 2 different null values are ok to interchange.  Still, i'll fix it, it's ugly

ezyang commented on caffe2/CMakeLists.txt:
```diff
@@ -407,8 +409,15 @@ if(NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
       $<$<BOOL:${INTERN_DISABLE_AUTOGRAD}>:--disable-autograd>
       $<$<BOOL:${SELECTED_OP_LIST}>:--selected-op-list-path="${SELECTED_OP_LIST}">
       --force_schema_registration
+      --gen_lazy_ts_backend
```
Any reason this isn't just true by default?

wconstab replied:
yea, I can more easily avoid solving all the build systems at once this way; it lets aros/ovrsource/xplat versions of generate_code continue doing what they used to do for now

ezyang commented on tools/codegen/gen_lazy_tensor.py:
```diff
@@ -236,22 +256,19 @@ def gen_key(func: FunctionSchema) -> Tuple[str, str]:
     })
 
     # Generate IR node classes
-    fm.write_with_template(f'{backend_key}LazyIr.h', 'LazyIr.h', lambda: {
```
this was just a straight bug before right?

wconstab commented on torch/csrc/lazy/core/hash.cpp:
```diff
@@ -89,5 +89,23 @@ std::string HashToString(const hash_t& a) {
+  // We can't assume a DataHash size/dataptr approach here bc
+  // vector<bool> can be optimized as vector<bit> and storage details
+  // are decoupled from actual size of 'bool' type
+  hash_t h = 0xad2ed1983bbf2e28;
```
@ezyang Wdyt about this? I had a failure in macos build, 

Compiling LazyIr.h is the first time we've exercised Hash(vector<bool) 

I also added a new test (see test_misc.cpp in this diff)

ezyang replied:
uhhh, sure! :) I'm not even sure what you find objectionable about this haha

wconstab committed:
Run lazy tensor codegen in generate_code.py (#73996)

Summary:
Hooks into existing autograd codegen script (generate_code.py) to take advantage of its integrations into buck/cmake/bazel.

Adds a new option (--gen_lazy_ts_backend) to. generate_code.py, calling this from CMake OSS build and fbcode build, but not from other internal xplat/ovrsource builds (these could be opted in later)

Bazel support is added in a later diff.

Includes one generated file (torch/csrc/lazy/generated/LazyIr.h) in a unit test (test/cpp/lazy/test_ir.cpp) to partially verify the generator is working, but does not compile the remaining output sources from the generator yet as they depend on other files not yet landed from lazy_tensor_staging branch.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/73996

Test Plan: OSS/internal CI - verify all builds are working and test_ir.cpp compiles LazyIr.h

Reviewed By: ezyang

Differential Revision: D34408536

fbshipit-source-id: b7d46d817b3ed3c56108d65bbf052ed73b4e0827
